cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/SortedReplicasHelper.java

```diff
@@ -0,0 +1,151 @@
+  private List<Function<Replica, Integer>> _priorityFuncs;
+  private Function<Replica, Double> _scoreFunc;
+
+  public SortedReplicasHelper() {
+    _selectionFuncs = new HashSet<>();
+    _priorityFuncs = new ArrayList<>();
+  }
+
+  /**
+   * If condition hold, add a selection function to the {@link SortedReplicas} to be created.
+   *
+   * @param selectionFunc The selection function to add.
+   * @param condition Whether add the selection function or not.
+   * @return The helper object itself.
+   */
+  public SortedReplicasHelper maybeAddSelectionFunc(Function<Replica, Boolean> selectionFunc, boolean condition) {
+    if (condition) {
+      _selectionFuncs.add(selectionFunc);
+    }
+    return this;
+  }
+
+  /**
+   * Add a selection function to the {@link SortedReplicas} to be created.
+   *
+   * @param selectionFunc The selection function to add.
+   * @return The helper object itself.
+   */
+  public SortedReplicasHelper addSelectionFunc(Function<Replica, Boolean> selectionFunc) {
+    return maybeAddSelectionFunc(selectionFunc, true);
+  }
+
+
+  /**
+   * If condition hold, add a priority function to the {@link SortedReplicas} to be created.
+   * Note the order adding priority function matters. Each call of this method or {@link this#addPriorityFunc(Function)}
+   * will add new priority function to the end of priority function list, which will be applied only when all the previous
+   * priority functions cannot resolve two replica's priority.
+   *
+   * @param priorityFunc The priority function to add.
+   * @param condition Whether add the priority function or not.
+   * @return The helper object itself.
+   */
+  public SortedReplicasHelper maybeAddPriorityFunc(Function<Replica, Integer> priorityFunc, boolean condition) {
+    if (condition && !_priorityFuncs.contains(priorityFunc)) {
+      _priorityFuncs.add(priorityFunc);
+    }
+    return this;
+  }
+
+  /**
+   * Add a priority function to the {@link SortedReplicas} to be created.
+   * Note the order adding priority function matters. Each call of this method or {@link this#maybeAddPriorityFunc(Function, boolean)}
+   * will add new priority function to the end of priority function list, which will be applied only when all the previous
+   * priority functions cannot resolve two replica's priority.
+   *
+   * @param priorityFunc The priority function to add.
+   * @return The helper object itself.
+   */
+  public SortedReplicasHelper addPriorityFunc(Function<Replica, Integer> priorityFunc) {
+    return maybeAddPriorityFunc(priorityFunc, true);
+  }
+
+  /**
+   * If condition hold, add a score function to the {@link SortedReplicas} to be created.
+   *
+   * @param scoreFunc The score function to add.
+   * @param condition Whether add the score function or not.
+   * @return The helper object itself.
+   */
+  public SortedReplicasHelper maybeAddScoreFunc(Function<Replica, Double> scoreFunc, boolean condition) {
```

Comment:
(Also applies to other JavaDocs) `condition` -> can we use more descriptive names? Also this function seems redundant, would you clarify its expected use-case?

Reply:
The main reason I want to have these `maybeAddxx` methods are to make the API more "confluent".
Consider [use case at CapacityGoal](https://github.com/linkedin/cruise-control/blob/d046302370e20c3bc780d09afabe2bf9342ede19/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/CapacityGoal.java#L182), without this API, I need to write 
```   
SortedReplicaHelper helper = new SortedReplicasHelper();
if (onlyMoveImmigrantReplicas) {
    helper.addSelectionFunc(ReplicaSortFunctionFactory.selectImmigrants();
}                     
 helper.addSelectionFunc(ReplicaSortFunctionFactory.selectReplicasNotFromExcludedTopics(excludedTopics))
           .addPriorityFunc(ReplicaSortFunctionFactory.prioritizeOfflineReplicas())
           .addPriorityFunc(ReplicaSortFunctionFactory.prioritizeImmigrants())
           ...
```

Reply:
you are referring only to the score function,  you are right. it it not used, will remove it.